Assets/Code/Scripts/RealityMovement.cs

```diff
@@ -1,53 +1,195 @@
+    {
+        Walking,
+        Sprinting,
+        Crouching,
+        Air
+    }
+    
+    [Header("Speed")]
+    [SerializeField] private float _walkSpeed = 6f;
+    [SerializeField] private float _sprintSpeed = 10f;
+    
+    private float _moveSpeed;     // speed intensity
+    
+    [Header("Drag")]
+    [SerializeField] private float _groundDrag = 4f;    // ground drag
+    
+    [Header("Jump")]
+    [SerializeField] private float _jumpForce = 8f;     // set jump upward force
+    [SerializeField] private float _jumpCooldown = 0.25f;      // set jump cooldown
+    [SerializeField] private float _airMultiplier = 0.4f;     // set air movement limitation
+    private bool _readyToJump;      //
 
-    [SerializeField] private Transform _groundCheck;
+    [Header("Crouch")]
+    [SerializeField] private float _crouchSpeed = 2f;
+    [SerializeField] private float _crouchYScale = 0.5f;
+    private float _startYScale;
     
-    [SerializeField] private LayerMask _groundMask;
+    [Header("Keybinds")]
+    [SerializeField] private KeyCode jumpKey = KeyCode.Space;     // set the jump key
+    [SerializeField] private KeyCode sprintKey = KeyCode.LeftShift;   // set the sprint key
+    [SerializeField] private KeyCode crouchKey = KeyCode.LeftControl;
+
+    [Header("Ground Check")]
+    // set the ground check
+    [SerializeField] private float _playerHeight = 2f;
+    [SerializeField] private LayerMask _ground;
+    private bool _grounded;
     
-    private Vector3 _velocity;
+    [SerializeField] private Transform _orientation;
+
+    private float _horizontalInput;
+    private float _verticalInput;
+
+    private Vector3 _moveDirection;
+
     private Transform _transform;
+    private Rigidbody _rigidbody;      // set the rigidbody
+
+    // player states
+    [SerializeField] private MovementState _state;     // current player state
 
     private void Awake()
     {
+        _rigidbody = GetComponent<Rigidbody>();
         _transform = GetComponent<Transform>();
     }
 
+    // Start is called before the first frame update
+    private void Start()
+    {
+        _rigidbody.freezeRotation = true;
+        
+        ResetJump(); // set _readyToJump to "true"
+
+        _startYScale = _transform.localScale.y;
+    }
+
     // Update is called once per frame
-    void Update()
+    private void Update()
     {
-        bool isGrounded = Physics.CheckSphere(_groundCheck.position, _groundDistance, _groundMask);
+        _grounded = Physics.Raycast(_transform.position, Vector3.down, _playerHeight * 0.5f + 0.2f, _ground);
+
+        MyInput();
+        StateHandler();
+        SpeedControl();
 
-        if (isGrounded && _velocity.y < 0)
+        // handle drag
+        if (_grounded)
+            _rigidbody.drag = _groundDrag;
+        else
+            _rigidbody.drag = 0;
+    }
+
+    private void FixedUpdate()
+    {
+        MovePlayer();
+        //Debug.Log(_moveSpeed);
+    }
+
+    // 
+    private void MyInput()
```

Comment:
Picky me: I would rename this function and/or add a small docstring to explain what it does, for me it is not 100% clear